src/Model/Behavior/ProfferBehavior.php

```diff
@@ -79,53 +79,131 @@ public function beforeMarshal(Event $event, ArrayObject $data, ArrayObject $opti
+     *
+     * @param string $field The upload field name
+     * @param array $settings Array of upload settings for the field
+     * @param \Cake\Datasource\EntityInterface $entity The current entity to process
+     * @param \Proffer\Lib\ProfferPathInterface|null $path Inject an instance of ProfferPath
+     *
+     * @throws \Exception If the file cannot be renamed / moved to the correct path
+     *
+     * @return void
+     */
+    protected function process($field, array $settings, EntityInterface $entity, ProfferPathInterface $path = null)
+    {
+        $path = $this->createPath($entity, $field, $settings, $path);
+        $tableEntityClass = $this->_table->entityClass();
 
-                    // Only generate thumbnails for image uploads
-                    if (getimagesize($path->fullPath()) !== false && isset($settings['thumbnailSizes'])) {
-                        // Allow the transformation class to be injected
-                        if (!empty($settings['transformClass'])) {
-                            $imageTransform = new $settings['transformClass']($this->_table, $path);
-                        } else {
-                            $imageTransform = new ImageTransform($this->_table, $path);
-                        }
+        if ($tableEntityClass !== null && $entity instanceof $tableEntityClass) {
+            $uploadList = [
+                [
+                    'name' => $entity->get('name'),
+                    'type' => $entity->get('type'),
+                    'tmp_name' => $entity->get('tmp_name'),
+                    'error' => $entity->get('error'),
+                    'size' => $entity->get('size')
+                ]
+            ];
+        } else {
+            $uploadList = $entity->get($field);
+            if (count(array_filter(array_keys($entity->get($field)), 'is_string')) > 0) {
+                $uploadList = [$entity->get($field)];
+            }
+        }
 
-                        $thumbnailPaths = $imageTransform->processThumbnails($settings);
-                        $imagePaths = array_merge($imagePaths, $thumbnailPaths);
+        foreach ($uploadList as $upload) {
+            if ($this->moveUploadedFile($upload['tmp_name'], $path->fullPath())) {
+                $entity->set($field, $path->getFilename());
+                $entity->set($settings['dir'], $path->getSeed());
 
-                        $eventData = ['path' => $path, 'images' => $imagePaths];
-                        $event = new Event('Proffer.afterCreateImage', $entity, $eventData);
-                        $this->_table->eventManager()->dispatch($event);
-                    }
-                } else {
-                    throw new Exception('Cannot upload file');
-                }
+                $this->createThumbnails($entity, $settings, $path);
+            } else {
+                throw new Exception('Cannot upload file');
```

Comment:
Ideally throw specific exception types that aid in describing the exception (e.g. `CannotUploadFileException`) - failing that, at minimum use SPL exceptions